File: sdk/go/checkpoint_test.go
package agenttrace

import (
	"context"
	"os"
	"path/filepath"
	"testing"
	"time"
)

func TestCheckpointType(t *testing.T) {
	t.Run("all checkpoint types exist", func(t *testing.T) {
		types := []CheckpointType{
			CheckpointTypeManual,
			CheckpointTypeAuto,
			CheckpointTypeToolCall,
			CheckpointTypeError,
			CheckpointTypeMilestone,
			CheckpointTypeRestore,
		}

		expected := []string{"manual", "auto", "tool_call", "error", "milestone", "restore"}

		for i, cp := range types {
			if string(cp) != expected[i] {
				t.Errorf("expected %s, got %s", expected[i], cp)
			}
		}
	})
}

func TestTraceCheckpoint(t *testing.T) {
	client := New(Config{
		APIKey: "test-api-key",
	})
	defer client.Shutdown()

	ctx := context.Background()
	trace := client.Trace(ctx, TraceOptions{
		Name: "test-trace",
	})

	t.Run("creates checkpoint with minimal options", func(t *testing.T) {
		cp := trace.Checkpoint(CheckpointOptions{
			Name: "test-checkpoint",
		})

		if cp.Name != "test-checkpoint" {
			t.Errorf("expected name to be 'test-checkpoint', got '%s'", cp.Name)
		}
		if cp.Type != CheckpointTypeManual {
			t.Errorf("expected type to be 'manual', got '%s'", cp.Type)
		}
		if cp.TraceID != trace.id {
			t.Error("expected checkpoint to have trace's ID")
		}
		if cp.ID == "" {
			t.Error("expected checkpoint ID to be generated")
		}
		if cp.CreatedAt.IsZero() {
			t.Error("expected CreatedAt to be set")
		}
	})

	t.Run("creates checkpoint with custom type", func(t *testing.T) {
		cp := trace.Checkpoint(CheckpointOptions{
			Name: "milestone-checkpoint",
			Type: CheckpointTypeMilestone,
		})

		if cp.Type != CheckpointTypeMilestone {
			t.Errorf("expected type to be 'milestone', got '%s'", cp.Type)
		}
	})

	t.Run("creates checkpoint with observation ID", func(t *testing.T) {
		cp := trace.Checkpoint(CheckpointOptions{
			Name:          "obs-checkpoint",
			ObservationID: "obs-456",
		})

		if cp.ObservationID != "obs-456" {
			t.Errorf("expected observationID to be 'obs-456', got '%s'", cp.ObservationID)
		}
	})

	t.Run("creates checkpoint with files", func(t *testing.T) {
		// Create temp file
		tmpDir := t.TempDir()
		tmpFile := filepath.Join(tmpDir, "test.go")
		if err := os.WriteFile(tmpFile, []byte("package main"), 0644); err != nil {
			t.Fatal(err)
		}

		cp := trace.Checkpoint(CheckpointOptions{
			Name:           "file-checkpoint",
			Files:          []string{tmpFile},
			IncludeGitInfo: false,
		})

		if cp.TotalFiles != 1 {
			t.Errorf("expected TotalFiles to be 1, got %d", cp.TotalFiles)
		}
		if cp.TotalSizeBytes == 0 {
			t.Error("expected TotalSizeBytes to be > 0")
		}
		if len(cp.FilesChanged) != 1 {
			t.Errorf("expected 1 file in FilesChanged, got %d", len(cp.FilesChanged))
		}
	})

	t.Run("handles nonexistent files gracefully", func(t *testing.T) {
		cp := trace.Checkpoint(CheckpointOptions{
			Name:           "no-file-checkpoint",
			Files:          []string{"/nonexistent/file.go"},
			IncludeGitInfo: false,
		})

		if cp.TotalFiles != 1 {
			t.Errorf("expected TotalFiles to be 1, got %d", cp.TotalFiles)
		}
		if cp.TotalSizeBytes != 0 {
			t.Errorf("expected TotalSizeBytes to be 0 for nonexistent file, got %d", cp.TotalSizeBytes)
		}
	})

	t.Run("creates checkpoint with empty files list", func(t *testing.T) {
		cp := trace.Checkpoint(CheckpointOptions{
			Name:           "empty-files-checkpoint",
			Files:          []string{},
			IncludeGitInfo: false,
		})

		if cp.TotalFiles != 0 {
			t.Errorf("expected TotalFiles to be 0, got %d", cp.TotalFiles)
		}
		if len(cp.FilesChanged) != 0 {
			t.Errorf("expected empty FilesChanged, got %d files", len(cp.FilesChanged))
		}
	})
}

func TestCheckpointDisabledClient(t *testing.T) {
	enabled := false
	client := New(Config{
		APIKey:  "test-api-key",
		Enabled: &enabled,
	})
	defer client.Shutdown()

	ctx := context.Background()
	trace := client.Trace(ctx, TraceOptions{
		Name: "test-trace",
	})

	cp := trace.Checkpoint(CheckpointOptions{
		Name: "disabled-checkpoint",
	})

	// Should still return checkpoint info
	if cp.Name != "disabled-checkpoint" {
		t.Errorf("expected checkpoint name even when disabled")
	}
	if cp.ID == "" {
		t.Error("expected checkpoint ID even when disabled")
	}
}

func TestCheckpointInfo(t *testing.T) {
	t.Run("has all expected fields", func(t *testing.T) {
		info := CheckpointInfo{
			ID:             "cp-123",
			Name:           "test",
			Type:           CheckpointTypeManual,
			TraceID:        "trace-456",
			ObservationID:  "obs-789",
			GitCommitSha:   "abc123",
			GitBranch:      "main",
			FilesChanged:   []string{"file1.go", "file2.go"},
			TotalFiles:     2,
			TotalSizeBytes: 1024,
			CreatedAt:      time.Now(),
		}

		if info.ID != "cp-123" {
			t.Error("ID not set correctly")
		}
		if info.Name != "test" {
			t.Error("Name not set correctly")
		}
		if info.Type != CheckpointTypeManual {
			t.Error("Type not set correctly")
		}
		if info.TraceID != "trace-456" {
			t.Error("TraceID not set correctly")
		}
		if info.ObservationID != "obs-789" {
			t.Error("ObservationID not set correctly")
		}
		if info.GitCommitSha != "abc123" {
			t.Error("GitCommitSha not set correctly")
		}
		if info.GitBranch != "main" {
			t.Error("GitBranch not set correctly")
		}
		if len(info.FilesChanged) != 2 {
			t.Error("FilesChanged not set correctly")
		}
		if info.TotalFiles != 2 {
			t.Error("TotalFiles not set correctly")
		}
		if info.TotalSizeBytes != 1024 {
			t.Error("TotalSizeBytes not set correctly")
		}
		if info.CreatedAt.IsZero() {
			t.Error("CreatedAt not set correctly")
		}
	})
}

func TestGetGitInfo(t *testing.T) {
	// This test will behave differently depending on whether
	// we're running in a git repo or not
	info := getGitInfo()

	// Just verify it doesn't panic and returns a GitInfo struct
	_ = info.CommitSha
	_ = info.Branch
	_ = info.RepoURL
}

func TestGitInfo(t *testing.T) {
	t.Run("struct holds git information", func(t *testing.T) {
		info := GitInfo{
			CommitSha: "abc123def456",
			Branch:    "main",
			RepoURL:   "https://github.com/test/repo.git",
		}

		if info.CommitSha != "abc123def456" {
			t.Error("CommitSha not set correctly")
		}
		if info.Branch != "main" {
			t.Error("Branch not set correctly")
		}
		if info.RepoURL != "https://github.com/test/repo.git" {
			t.Error("RepoURL not set correctly")
		}
	})
}

File: sdk/typescript/tests/checkpoint.test.ts
/**
 * Tests for the AgentTrace TypeScript SDK checkpoint module.
 */

import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { existsSync, statSync, readFileSync } from "fs";
import { execSync } from "child_process";

// Mock fs and child_process
vi.mock("fs", () => ({
  existsSync: vi.fn(),
  statSync: vi.fn(),
  readFileSync: vi.fn(),
}));

vi.mock("child_process", () => ({
  execSync: vi.fn(),
}));

vi.mock("crypto", () => ({
  createHash: vi.fn(() => ({
    update: vi.fn().mockReturnThis(),
    digest: vi.fn().mockReturnValue("abc123hash"),
  })),
}));

import { CheckpointClient, CheckpointInfo, CheckpointOptions, withCheckpoint } from "../src/checkpoint";

describe("CheckpointClient", () => {
  let mockClient: any;
  let checkpointClient: CheckpointClient;

  beforeEach(() => {
    mockClient = {
      enabled: true,
      _addEvent: vi.fn(),
    };
    checkpointClient = new CheckpointClient(mockClient);
    vi.clearAllMocks();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe("create", () => {
    it("should create a checkpoint with minimal options", () => {
      const cp = checkpointClient.create("trace-123", {
        name: "test-checkpoint",
      });

      expect(cp.name).toBe("test-checkpoint");
      expect(cp.type).toBe("manual");
      expect(cp.traceId).toBe("trace-123");
      expect(cp.id).toBeDefined();
      expect(cp.createdAt).toBeInstanceOf(Date);
    });

    it("should create a checkpoint with custom type", () => {
      const cp = checkpointClient.create("trace-123", {
        name: "milestone-cp",
        type: "milestone",
      });

      expect(cp.type).toBe("milestone");
    });

    it("should create a checkpoint with observation ID", () => {
      const cp = checkpointClient.create("trace-123", {
        name: "obs-checkpoint",
        observationId: "obs-456",
      });

      expect(cp.observationId).toBe("obs-456");
    });

    it("should track files when provided", () => {
      vi.mocked(existsSync).mockReturnValue(true);
      vi.mocked(statSync).mockReturnValue({
        size: 1024,
      } as any);
      vi.mocked(readFileSync).mockReturnValue(Buffer.from("test content"));

      const cp = checkpointClient.create("trace-123", {
        name: "file-checkpoint",
        files: ["/path/to/file1.ts", "/path/to/file2.ts"],
      });

      expect(cp.totalFiles).toBe(2);
      expect(cp.filesChanged).toEqual(["/path/to/file1.ts", "/path/to/file2.ts"]);
    });

    it("should calculate total size from files", () => {
      vi.mocked(existsSync).mockReturnValue(true);
      vi.mocked(statSync).mockReturnValue({
        size: 500,
      } as any);
      vi.mocked(readFileSync).mockReturnValue(Buffer.from("test"));

      const cp = checkpointClient.create("trace-123", {
        name: "size-checkpoint",
        files: ["/path/file1.ts", "/path/file2.ts"],
      });

      expect(cp.totalSizeBytes).toBe(1000); // 500 + 500
    });

    it("should handle nonexistent files gracefully", () => {
      vi.mocked(existsSync).mockReturnValue(false);

      const cp = checkpointClient.create("trace-123", {
        name: "no-files-checkpoint",
        files: ["/nonexistent/file.ts"],
      });

      expect(cp.totalFiles).toBe(1);
      expect(cp.totalSizeBytes).toBe(0);
    });

    it("should send event to API when client is enabled", () => {
      const cp = checkpointClient.create("trace-123", {
        name: "api-checkpoint",
        type: "auto",
      });

      expect(mockClient._addEvent).toHaveBeenCalled();
      const eventArg = mockClient._addEvent.mock.calls[0][0];
      expect(eventArg.type).toBe("checkpoint-create");
      expect(eventArg.body.traceId).toBe("trace-123");
      expect(eventArg.body.name).toBe("api-checkpoint");
      expect(eventArg.body.type).toBe("auto");
    });

    it("should not send event when client is disabled", () => {
      mockClient.enabled = false;

      const cp = checkpointClient.create("trace-123", {
        name: "disabled-checkpoint",
      });

      expect(mockClient._addEvent).not.toHaveBeenCalled();
      expect(cp.name).toBe("disabled-checkpoint"); // Still returns info
    });
  });

  describe("getGitInfo", () => {
    it("should get git commit sha", () => {
      vi.mocked(execSync).mockImplementation((cmd: string) => {
        if (cmd.includes("rev-parse HEAD")) {
          return "abc123def456\n";
        }
        return "";
      });

      const gitInfo = checkpointClient.getGitInfo();

      expect(gitInfo.commitSha).toBe("abc123def456");
    });

    it("should get git branch", () => {
      vi.mocked(execSync).mockImplementation((cmd: string) => {
        if (cmd.includes("--abbrev-ref HEAD")) {
          return "feature-branch\n";
        }
        return "";
      });

      const gitInfo = checkpointClient.getGitInfo();

      expect(gitInfo.branch).toBe("feature-branch");
    });

    it("should get repo URL", () => {
      vi.mocked(execSync).mockImplementation((cmd: string) => {
        if (cmd.includes("remote.origin.url")) {
          return "https://github.com/test/repo.git\n";
        }
        return "";
      });

      const gitInfo = checkpointClient.getGitInfo();

      expect(gitInfo.repoUrl).toBe("https://github.com/test/repo.git");
    });

    it("should handle git command failures gracefully", () => {
      vi.mocked(execSync).mockImplementation(() => {
        throw new Error("git not found");
      });

      const gitInfo = checkpointClient.getGitInfo();

      expect(gitInfo.commitSha).toBeUndefined();
      expect(gitInfo.branch).toBeUndefined();
      expect(gitInfo.repoUrl).toBeUndefined();
    });
  });

  describe("checkpoint with git info", () => {
    it("should include git info by default", () => {
      vi.mocked(execSync).mockImplementation((cmd: string) => {
        if (cmd.includes("rev-parse HEAD")) return "commit123\n";
        if (cmd.includes("--abbrev-ref")) return "main\n";
        return "";
      });

      const cp = checkpointClient.create("trace-123", {
        name: "git-checkpoint",
        includeGitInfo: true,
      });

      expect(cp.gitCommitSha).toBe("commit123");
      expect(cp.gitBranch).toBe("main");
    });

    it("should exclude git info when requested", () => {
      vi.mocked(execSync).mockReturnValue("should-not-appear\n");

      const cp = checkpointClient.create("trace-123", {
        name: "no-git-checkpoint",
        includeGitInfo: false,
      });

      expect(cp.gitCommitSha).toBeUndefined();
      expect(cp.gitBranch).toBeUndefined();
    });
  });
});

describe("CheckpointType", () => {
  it("should support all checkpoint types", () => {
    const types: Array<"manual" | "auto" | "tool_call" | "error" | "milestone" | "restore"> = [
      "manual",
      "auto",
      "tool_call",
      "error",
      "milestone",
      "restore",
    ];

    const mockClient = {
      enabled: true,
      _addEvent: vi.fn(),
    };
    const client = new CheckpointClient(mockClient);

    for (const type of types) {
      const cp = client.create("trace-123", {
        name: `${type}-checkpoint`,
        type,
      });
      expect(cp.type).toBe(type);
    }
  });
});

describe("withCheckpoint", () => {
  it("should create checkpoint and execute function", async () => {
    const mockClient = {
      enabled: true,
      _addEvent: vi.fn(),
    };

    const result = await withCheckpoint(
      mockClient as any,
      "trace-123",
      { name: "scope-checkpoint" },
      async (cp) => {
        expect(cp.name).toBe("scope-checkpoint");
        return "result-value";
      }
    );

    expect(result).toBe("result-value");
  });

  it("should propagate errors from function", async () => {
    const mockClient = {
      enabled: true,
      _addEvent: vi.fn(),
    };

    await expect(
      withCheckpoint(
        mockClient as any,
        "trace-123",
        { name: "error-checkpoint" },
        async () => {
          throw new Error("Test error");
        }
      )
    ).rejects.toThrow("Test error");
  });
});

describe("CheckpointInfo", () => {
  it("should have all expected fields", () => {
    const mockClient = {
      enabled: false,
      _addEvent: vi.fn(),
    };
    const client = new CheckpointClient(mockClient);

    const cp = client.create("trace-123", {
      name: "full-checkpoint",
      type: "milestone",
      observationId: "obs-456",
      description: "Test description",
      files: [],
    });

    expect(cp).toHaveProperty("id");
    expect(cp).toHaveProperty("name");
    expect(cp).toHaveProperty("type");
    expect(cp).toHaveProperty("traceId");
    expect(cp).toHaveProperty("observationId");
    expect(cp).toHaveProperty("filesChanged");
    expect(cp).toHaveProperty("totalFiles");
    expect(cp).toHaveProperty("totalSizeBytes");
    expect(cp).toHaveProperty("createdAt");
  });
});
